doc(pe): Document policy options (#601)
armory/docs: content/en/docs/plugin-guide/plugin-policy-engine/packages/spinnaker.execution/stages.before/cloneServerGroup.md

---
title: "spinnaker.execution.stages.before.cloneServerGroup"
linktitle: "cloneServerGroup"

> description: "A policy targeting this object is run before executing each task in a cloneServerGroup stage."

Reviewer: ```suggestion
description: "A policy targeting this object runs before executing each task in a cloneServerGroup stage."
```

## Example Policy

{{< prism lang="rego" line-numbers="true" >}}
package spinnaker.execution.stages.before.cloneServerGroup

productionAccounts :=["prod1","prod2"]

deny["ASGs running in production must have a minimum of 2 instances to avoid having a single point of failure."]{
	input.stage.context.credentials==productionAccounts[_]
    input.stage.context.useSourceCapacity==false
    object.get(input.stage.context.capacity,"min",0)<2
}{
	input.stage.context.credentials==productionAccounts[_]
    input.stage.context.useSourceCapacity==false
    object.get(input.stage.context,"capacity",null)==null
}
{{< /prism >}}

## Keys

Parameters related to the stage against which the policy is executing can be found in the [input.stage.context](#inputstagecontext) object.

### input.pipeline

| Key                                               | Type      | Description                                                                                                                           |
| ------------------------------------------------- | --------- | ------------------------------------------------------------------------------------------------------------------------------------- |

> | `input.pipeline.application`                      | `string`  | The name of the Spinnaker application to which this pipeline belongs.                                                                 |

Reviewer: ```suggestion
| `input.pipeline.application`                      | `string`  | The name of the application to which this pipeline belongs.                                                                 |
```

| `input.pipeline.authentication.allowedAccounts[]` | `string`  | The list of accounts to which the user this stage is running as has access.                                                           |

> | `input.pipeline.authentication.user`              | `string`  | The Spinnaker user initiating the change.                                                                                             |

Reviewer: ```suggestion
| `input.pipeline.authentication.user`              | `string`  | The user initiating the change.                                                                                             |
```

| `input.pipeline.origin`                           | `string`  |                                                                                                                                       |
| `input.pipeline.partition`                        |           |                                                                                                                                       |
| `input.pipeline.paused`                           |           |                                                                                                                                       |
| `input.pipeline.pipelineConfigId`                 | `string`  |                                                                                                                                       |
| `input.pipeline.source`                           |           |                                                                                                                                       |

> | `input.pipeline.spelEvaluator`                    | `string`  | Which version of spring expression language is being used to evaluate SpEL.                                                           |

Reviewer: ```suggestion
| `input.pipeline.spelEvaluator`                    | `string`  | Which version of Spring Expression Language is being used to evaluate SpEL expressions.                                                           |
```

Reviewer: Find/replace them all probably

> | `input.pipeline.stages[]`                         | `[array]` | An array of the stages in the pipeline. Typically if you are writing a policy that examines multiple pipeline stages, it is better to write that policy against either the `opa.pipelines package`, or the `spinnaker.execution.pipelines.before` package. |

Reviewer: ```suggestion
| `input.pipeline.stages[]`                         | `[array]` | An array of the stages in the pipeline. Typically if you are writing a policy that examines multiple pipeline stages, it is better to write that policy against either the `opa.pipelines package` or the `spinnaker.execution.pipelines.before` package. |
```

| `input.pipeline.status`                           | `string`  |                                                                                                                                       |
| `input.pipeline.templateVariables`                |           |                                                                                                                                       |
| `input.pipeline.type`                             | `string`  |                                                                                                                                       |

### input.pipeline.trigger

See [input.pipeline.trigger]({{< ref "input.pipeline.trigger.md" >}}) for more information.

### input.stage

See [`input.stage`]({{< ref "input.stage.md" >}}) for more information.

### input.stage.context

| Key                                                       | Type      | Description                                                                                           |
| --------------------------------------------------------- | --------- | ----------------------------------------------------------------------------------------------------- |

> | `input.stage.context.application`                         | `string`  | The name of the Spinnaker application for this pipeline.                                              |

Reviewer: ```suggestion
| `input.stage.context.application`                         | `string`  | The name of the application for this pipeline.                                              |
```

> | `input.stage.context.attempt`                             | `number`  | How many times has this stage attempted to execute.                                                   |

Reviewer: This one could use some additional info

| `input.stage.context.capacity.desired`                    | `number`  | If `useSourceCapacity` is false, the desired number of instances to run in the group.                 |
| `input.stage.context.capacity.max`                        | `number`  | If `useSourceCapacity` is false, the maximum number of instances to run in the group.                 |
| `input.stage.context.capacity.min`                        | `number`  | If `useSourceCapacity` is false, the minimum number of instances to run in the group.                 |
| `input.stage.context.cloudProvider`                       | `string`  | The name of the cloud provider that will execute the stage.                                           |

> | `input.stage.context.cloudProviderType`                   | `string`  | The type of the cloud provider that will execute the stage.                                           |

Reviewer: What're the types?

| `input.stage.context.freeFormDetails`                     | `string`  |                                                                                                       |
| `input.stage.context.lastException`                       | `string`  | The last exception that occurred when executing this stage.                                           |
| `input.stage.context.region`                              | `string`  | The region in which the new server group will be deployed.                                            |
| `input.stage.context.stack`                               | `string`  |                                                                                                       |
| `input.stage.context.target`                              | `string`  | Which server group should be selected when this stage starts.                                         |
| `input.stage.context.targetCluster`                       | `string`  | The cluster that contains the target.                                                                 |

> | `input.stage.context.useAmiBlockDeviceMappings`           | `boolean` | Spinnaker will use the block device mappings from the selected AMI when deploying a new server group. |

Reviewer: ```suggestion
| `input.stage.context.useAmiBlockDeviceMappings`           | `boolean` | If true, Armory Enterprise uses the block device mappings from the selected AMI when deploying a new server group. |
```

> | `input.stage.context.useSourceCapacity`                   | `boolean` | Spinnaker will use the current capacity of the existing server group when deploying a new server group. This setting is intended to support a server group with auto-scaling enabled, where the bounds and desired capacity are controlled by an external process. In the event that there is no existing server group, the deploy will fail. |

Reviewer: ```suggestion
| `input.stage.context.useSourceCapacity`                   | `boolean` | Armory Enterprise uses the current capacity of the existing server group when deploying a new server group. This setting is intended to support a server group with auto-scaling enabled where the bounds and desired capacity are controlled by an external process. In the event that there is no existing server group, the deployment fails. |
```
